ddtrace/tracer.py

```diff
@@ -41,6 +53,59 @@
         logging.basicConfig(level=logging.DEBUG)
 
 
+@attr.s()
+class _Trace(object):
+    trace_id = attr.ib(type=int)  # type: int
+    sampled = attr.ib(type=bool, default=True)  # type: bool
+    num_finished = attr.ib(type=int, default=0)  # type: int
+    sampling_priority = attr.ib(default=None)  # type: Optional[int]
+    dd_origin = attr.ib(default=None)  # type: Optional[str]
+    local_root_span = attr.ib(default=None)  # type: Optional[Span]
+    _spans = attr.ib(default=attr.Factory(list))  # type: List[Span]
+
+    def __len__(self):
+        return len(self._spans)
+
+    def add_span(self, span):
+        # type: (Span) -> None
+        self._spans.append(span)
+
+    def pop_finished_spans(self):
+        # type: () -> List[Span]
+        local_root_span = self.local_root_span
+        if local_root_span:
+            if self.sampling_priority is not None and self.sampled:
+                local_root_span.set_metric(SAMPLING_PRIORITY_KEY, self.sampling_priority)
+            if self.dd_origin:
+                local_root_span.meta[ORIGIN_KEY] = str(self.dd_origin)
+
+        finished_spans = [span for span in self._spans if span.finished]
+        self._spans = [span for span in self._spans if not span.finished]
```

Comment:
There is no way for new spans to finish between these two statements right?

Would we be better off with:

```python
[span for span in self._spans if span not in finished_spans]
```

?

probably more expensive.

Also, can we do this effectively in one loop?

Reply:
It is possible for a span to finish between the two, yes, good point. 

Funny enough, doing two list comprehensions is faster that a single for loop building 2 lists.

![image](https://user-images.githubusercontent.com/6321485/106310200-32cac080-6231-11eb-9874-ca549b8ea745.png)
![image](https://user-images.githubusercontent.com/6321485/106310223-38280b00-6231-11eb-9cf2-9ba347bfd459.png)
![image](https://user-images.githubusercontent.com/6321485/106310236-3cecbf00-6231-11eb-8b93-097f3f05f55b.png)

Reply:
I'd be curious to know why, but if I had to guess there are probably some internal optimizations for building a PyList from a list comprehension rather than introducing the function call per-item.

what happens if we have in a single statement?

```python
finished_spans, self._spans = [s for s in self._spans if s.finished], [s for s in self._spans if not s.finished]
```

multiple assignment is atomic right?

Reply:
> multiple assignment is atomic right?

Naively I don't think so, but I'll look into it. I'm guessing they're gonna be different bytecode operations which could be interrupted.

Reply:
They are not atomic. You can use `dis` to show the bytecode operations:
```python
>>> def x():
...     finished_spans, self._spans = [s for s in self._spans if s.finished], [s for s in self._spans if not s.finished]
...
>>> dis.dis(x)
  2           0 LOAD_CONST               1 (<code object <listcomp> at 0x10301c710, file "<stdin>", line 2>)
              2 LOAD_CONST               2 ('x.<locals>.<listcomp>')
              4 MAKE_FUNCTION            0
              6 LOAD_GLOBAL              0 (self)
              8 LOAD_ATTR                1 (_spans)
             10 GET_ITER
             12 CALL_FUNCTION            1
             14 LOAD_CONST               3 (<code object <listcomp> at 0x10301c5b0, file "<stdin>", line 2>)
             16 LOAD_CONST               2 ('x.<locals>.<listcomp>')
             18 MAKE_FUNCTION            0
             20 LOAD_GLOBAL              0 (self)
             22 LOAD_ATTR                1 (_spans)
             24 GET_ITER
             26 CALL_FUNCTION            1
             28 ROT_TWO
             30 STORE_FAST               0 (finished_spans)
             32 LOAD_GLOBAL              0 (self)
             34 STORE_ATTR               1 (_spans)
             36 LOAD_CONST               0 (None)
             38 RETURN_VALUE

Disassembly of <code object <listcomp> at 0x10301c710, file "<stdin>", line 2>:
  2           0 BUILD_LIST               0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                14 (to 20)
              6 STORE_FAST               1 (s)
              8 LOAD_FAST                1 (s)
             10 LOAD_ATTR                0 (finished)
             12 POP_JUMP_IF_FALSE        4
             14 LOAD_FAST                1 (s)
             16 LIST_APPEND              2
             18 JUMP_ABSOLUTE            4
        >>   20 RETURN_VALUE

Disassembly of <code object <listcomp> at 0x10301c5b0, file "<stdin>", line 2>:
  2           0 BUILD_LIST               0
              2 LOAD_FAST                0 (.0)
        >>    4 FOR_ITER                14 (to 20)
              6 STORE_FAST               1 (s)
              8 LOAD_FAST                1 (s)
             10 LOAD_ATTR                0 (finished)
             12 POP_JUMP_IF_TRUE         4
             14 LOAD_FAST                1 (s)
             16 LIST_APPEND              2
             18 JUMP_ABSOLUTE            4
        >>   20 RETURN_VALUE
```

There's a lot of operations here that could be interrupted.